src/Query/SqlQueryAlter.php

```diff
@@ -0,0 +1,172 @@
+    }
+
+  }
+
+  /**
+   * Tries to ensure that a given table exists.
+   *
+   * @return string
+   *   The table name/alias used in the query.
+   */
+  protected static function ensureTable(Select $select, $table, EntityTypeInterface $entity_type) {
+    $tables = $select->getTables();
+    foreach ($tables as $table_info) {
+      if (isset($table_info['table']) && $table_info['table'] === $table) {
+        return $table_info['alias'];
+      }
+    }
+
+    // @todo Is this the right join?
+    return $select->innerJoin($table, NULL, $entity_type->getBaseTable() . '.' . $entity_type->getKey('id') . ' = ' . '%alias.entity_id');
+  }
+
+  /**
+   * Apply the entity conditions recursively to the SQL condition.
+   *
+   * @return \Drupal\Core\Database\Query\Condition
+   *   The modified SQL condition.
+   */
+  protected function applyCondition(EntityTypeInterface $entity_type, DefaultTableMapping $table_mapping, Select $select, SqlCondition $sql_condition, ConditionGroup $condition) {
+    $entity_type_id = $entity_type->id();
+    $field_storage_definitions = $this->entityFieldManager->getFieldStorageDefinitions($entity_type_id);
+
+    foreach ($condition->getConditions() as $cond) {
+      // Support nested conditions.
+      if ($cond instanceof ConditionGroup) {
+        $sql_condition->condition($this->applyCondition($entity_type, $table_mapping, $select, $sql_condition->conditionGroupFactory($cond->getConjunction()), $cond));
+      }
+      else {
+        $field_storage_definition = $field_storage_definitions[$cond->getField()];
+
+        // Determine the real table and column before applying the condition.
+        if ($field_storage_definition->isBaseField()) {
+          $table_name = $entity_type->getDataTable() ?: $entity_type->getBaseTable();
+        }
+        else {
+          $table_name = $table_mapping->getDedicatedDataTableName($field_storage_definition);
+        }
+
+        $table_alias = static::ensureTable($select, $table_name, $entity_type);
+
+        // @todo we need support for non main properties?
+        // @todo we need support for revision queries?
+
+        $field_column_name = $table_mapping->getFieldColumnName($field_storage_definition, $field_storage_definition->getMainPropertyName());
+
+        $sql_condition->condition("{$table_alias}.{$field_column_name}", $cond->getValue(), $cond->getOperator());
+      }
+    }
+
+    return $sql_condition;
+  }
+
+  /**
+   * Applies the cacheablity metadata to the current request.
+   *
+   * @param \Drupal\Core\Cache\CacheableMetadata $cacheable_metadata
+   *   The cacheability metadata.
+   */
+  protected function applyCacheability(CacheableMetadata $cacheable_metadata) {
+    $request = \Drupal::requestStack()->getCurrentRequest();
+    $renderer = \Drupal::service('renderer');
+    if ($request->isMethodCacheable() && $renderer->hasRenderContext()) {
+      $build = [];
+      $cacheable_metadata->applyTo($build);
+      $renderer->render($build);
```

Comment:
This bubbles onto the global render context just like we do in `\Drupal\Core\Render\MetadataBubblingUrlGenerator::bubble()`. But we did _that_ because we had to layer it on top after the fact. Ideally, you don't need this bubbling.

I suspect you need it here too because doing otherwise would break BC of the Entity Query API?

Reply:
This is dawehner's code, made to follow the core example. Can't say I understand it, so I kept it.